### PR TITLE
Update documentation to mention discrepancy between SDPA and Transformer/MHA mask definition

### PR DESCRIPTION
**Pull request by Rohan138**
Currently, `torch.nn.functional._canonical_mask` inverts the mask given input of type `torch.bool`, breaking all of the forward calls in `nn.Transformer`. This function seems to have been introduced in #92733.

Minimum repro example:
```
import torch
from torch.nn import functional as F

mask = torch.eye(2, 2, dtype=torch.bool)
print(mask)
canonincal_mask = F._canonical_mask(mask=mask, mask_name="mask", other_type=None, other_name=None, target_type=torch.float16)
print(canonincal_mask)
```
Output:
```
mask =
tensor([[ True, False],
        [False,  True]])
canonical_mask =
tensor([[-inf, 0.],
        [0., -inf]], dtype=torch.float16)
```

Expected:
```
tensor([[0., -inf],
        [-inf, 0.]], dtype=torch.float16)
```

Request: If someone could either add a test to prevent future regressions like this, or let me know where I could add one, that'd be great, thanks!

### CODE REVIEW

**Comment by Rohan138** [comment]
@jbschlosser @soumith Hi, sorry for the ping-could I get some assistance merging this? It's a minor bug but affects all masks in `nn.Transformer`.

The CI tests that are failing in `test/nn/test_multihead_attention` are also failing on main and seem unrelated.

Let me know if there's any other lint/bug fixes I'm missing.

**Review by mikaylagawarecki — CHANGES_REQUESTED** [review]
Hey, I do not think this is the correct change. Per the docs, `True` indicates that a position should be masked out. This is different from [`nn.functional.scaled_dot_product_attention`](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch-nn-functional-scaled-dot-product-attention) where the opposite is true (i.e. `True` indicates that a position should participate in the attention).

See the [note at the bottom of the docs](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html#torch.nn.Transformer.forward) for `torch.nn.Transformer`

> Note: [src/tgt/memory]_mask ensures that position i is allowed to attend the unmasked positions. If a BoolTensor is provided, positions with True are not allowed to attend while False values will be unchanged. If a FloatTensor is provided, it will be added to the attention weight. [src/tgt/memory]_key_padding_mask provides specified elements in the key to be ignored by the attention. If a BoolTensor is provided, the positions with the value of True will be ignored while the position with the value of False will be unchanged.

Similarly see [the docs](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) for `key_padding_mask` and `attn_mask` in  `nn.functional.MultiheadAttention`

> For a binary mask, a True value indicates that the corresponding key value will be ignored for the purpose of attention. For a float mask, it will be directly added to the corresponding key value.

**Comment by Rohan138** [comment]
@mikaylagawarecki ah okay, I see.

I can close this PR, but would it be possible to either use the same convention across both `nn` and `scaled_dot_product_attention`, or add a note highlighting the difference?

It's a little confusing as it is right now. Thanks!

**Comment by mikaylagawarecki** [comment]
@Rohan138 That sounds reasonable to me, would you be willing to open a PR that makes the suggested docs update? :)

**Comment by Rohan138** [comment]
Will do, thanks!

**Review by mikaylagawarecki — COMMENTED** [review]
Thanks! Could we document the discrepancy in the Transformer docs  and link to it in MHA instead please
